File: 卡码/代码随想录/1/02-链表/Python/07-环形链表/LeetCode_0142.py
#142. 环形链表 II
from typing import *

class ListNode:
	def __init__(self,val=0,next=None):
		self.val=val
		self.next=next

class Solution:
		def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
			slow,fast=head,head
			while fast!=None and fast.next!=None:
				slow=slow.next
				fast=fast.next.next
				#快慢指针相交必有环
				if slow==fast:
					index1,index2=fast,head
					while index1!=index2:
						index1,index2=index1.next,index2.next
					return index1
			return None

File: 卡码/代码随想录/1/02-链表/C++/07-环形链表/LeetCode_0142.cpp
//142. 环形链表 II
struct ListNode{
	int val;
	ListNode* next;
	ListNode():val(0),next(nullptr){}
	ListNode(int val):val(val),next(nullptr){}
	ListNode(int val,ListNode* next):val(val),next(next){}
};

class Solution {
public:
  ListNode *detectCycle(ListNode *head) {
  	ListNode *fast=head,*slow=head;
  	while(fast&&fast->next){
			slow=slow->next;
			fast=fast->next->next;
			//快慢指针相交必有环
			if(slow==fast){
				ListNode *index1=fast,*index2=head;
				while(index1!=index2){
					index1=index1->next;
					index2=index2->next;
				}
				return index1;
			}
		}
		return nullptr;
  }
};
